README.md
<!--
===============================================
vidgear library source-code is deployed under the Apache 2.0 License:

Copyright (c) 2019 Abhishek Thakur(@abhiTronix) <abhi.una12@gmail.com>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
===============================================
-->

<h1 align="center">
  <img src="https://abhitronix.github.io/img/vidgear/vidgear logo.svg" alt="VidGear Logo" width="70%"/>
</h1>
<h2 align="center">
  <img src="https://abhitronix.github.io/img/vidgear/vidgear banner.svg" alt="VidGear Banner" width="40%"/>
</h2>

<div align="center">

[Releases][release]&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;[Gears](#gears)&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;[Wiki Documentation][wiki]&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;[Installation](#installation)&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;[License](#license)

[![Build Status][travis-cli]][travis] [![Codecov branch][codecov]][code] [![Build Status][appveyor]][app]

[![Glitter chat][gitter-bagde]][gitter] [![PyPi version][pypi-badge]][pypi] [![Twitter][twitter-badge]][twitter-intent]

[![Code Style][black-badge]][black]

</div>

&nbsp;

VidGear is a powerful python Video Processing library built with multiple APIs *(a.k.a [**Gears**](#gears))* each with a unique set of trailblazing features. These APIs provides an easy-to-use, highly extensible, multi-threaded & asyncio wrapper around many underlying state-of-the-art libraries such as *[OpenCV ➶][opencv], [FFmpeg ➶][ffmpeg], [ZeroMQ ➶][zmq], [picamera ➶][picamera], [starlette ➶][starlette], [pafy ➶][pafy] and [python-mss ➶][mss]*

&nbsp;

The following **functional block diagram** clearly depicts the generalized functioning of VidGear library:

<p align="center">
  <img src="https://abhitronix.github.io/img/vidgear/vidgear_function2020_3.svg" alt="@Vidgear Functional Block Diagram" />
</p>

&nbsp;

# Table of Contents

* [**TL;DR**](#tldr)
* [**Gears: What are these?**](#gears)
  * [**CamGear**](#camgear)
  * [**PiGear**](#pigear)
  * [**VideoGear**](#videogear)
  * [**ScreenGear**](#screengear)
  * [**WriteGear**](#writegear)
  * [**NetGear**](#netgear)
  * [**WebGear**](#webgear)
  * [**NetGear_Async**](#netgear_async)
* [**New-Release SneekPeak: v0.1.7**](#new-release-sneekpeak--vidgear-017)
* [**Documentation**](#documentation)
* [**Installation**](#installation)
  * [**Prerequisites**](#prerequisites)
    * [**Supported Systems**](#supported-systems)
    * [**Supported Python legacies**](#supported-python-legacies)
    * [**Pip Dependencies**](#pip-dependencies)
  * [**Available Installation Options**](#available-installation-options)
    * [**PyPI Install**](#option-1-pypi-installrecommended)
    * [**Release Download**](#option-2-release-download)
    * [**Build from source**](#option-3-build-from-source)
* [**Testing, Formatting & Linting**](#testing-formatting--linting)
  * [**Requirements**](#requirements)
  * [**Running Tests**](#running-tests)
  * [**Formatting & Linting**](#formatting--linting)
* [**Contributions & Support**](#contributions--support)
  * [**Support**](#support)
  * [**Contributors**](#contributors)
* [**Community Channel**](#community-channel)
* [**Citing**](#citing)
* [**Copyright**](#copyright)


&nbsp;

&nbsp;



# TL;DR
  
#### What is vidgear?

> ***"VidGear is an [ultrafast➶][ultrafast-wiki], compact, flexible and easy-to-adapt complete Video Processing Python Library."***

#### What does it do?
> ***"VidGear can read, write, process, send & receive video frames from/to various devices in real-time."***

#### What is its purpose?
> ***"Built with simplicity in mind, VidGear lets programmers and software developers to easily integrate and perform complex Video Processing tasks in their existing or new applications, without going through various underlying library's documentation and using just a [few lines of code][flic]. Beneficial for both, if you're new to programming with Python language or already a pro at it."***

\
**For more information, see [*Frequently Asked Questions ➶*][faq].**


&nbsp;

&nbsp;


# Gears

> **VidGear is built with multiple APIs *(a.k.a Gears)*, each with some unique function/mechanism.**

Each of these APIs is exclusively designed to handle/control different device-specific video streams, network streams, and media encoders. These APIs provide an easy-to-use, highly extensible, multi-threaded and asyncio wrapper around state-of-the-art libraries under the hood to exploit their internal parameters and methods flexibly while providing robust error-handling and unparalleled performance. 

**These Gears can be classified as follows:**

**A. VideoCapture Gears:**

  * [**CamGear:**](#camgear) _Targets various IP-USB-Cameras/Network-Streams/YouTube-Video-URL._
  * [**PiGear:**](#pigear) _Targets various Raspberry Pi Camera Modules._
  * [**ScreenGear:**](#screengear) _Enables ultra-fast Screen Casting._    
  * [**VideoGear:**](#videogear) _Common API with Video Stabilizer wrapper._  

**B. VideoWriter Gear:**

  * [**WriteGear:**](#writegear) _Handles easy Lossless Video Encoding and Compression._

**C. Network Gears:**

  * [**NetGear:**](#netgear) _Targets flexible video-frames and data transfer between interconnecting systems over the network._

  * **Asynchronous I/O Network Gears:**

    * [**WebGear:**](#webgear) _ASGI Video Server that transfers live video frames to any web browser on the network._
    * [**NetGear_Async:**](#netgear_async) _Fast, Memory-Efficient Asyncio video-frame messaging framework._ 


&nbsp;

&nbsp;


## CamGear

> *CamGear can grab ultra-fast frames from diverse range of devices/streams, which includes almost any IP/USB Cameras, multimedia video file format ([_upto 4k tested_][test-4k]), various network stream protocols such as `http(s), rtp, rstp, rtmp, mms, etc.`, plus support for live Gstreamer's stream pipeline and YouTube video/live-streams URLs.*

CamGear provides a flexible, high-level multi-threaded wrapper around `OpenCV's` [VideoCapture class][opencv-vc] with access almost all of its available parameters and also employs [`pafy`][pafy] python APIs for live [YouTube streaming][youtube-wiki]. Furthermore, CamGear implements exclusively on [**Threaded Queue mode**][TQM-wiki] for ultra-fast, error-free and synchronized frame handling.


**Following functional block diagram depicts CamGear API's generalized workflow:**

<p align="center">
  <img src="https://github.com/abhiTronix/Imbakup/raw/master/Images/vidgear/camgearz2.png" alt="CamGear Functional Block Diagram"  width="70%"/>
</p>

### CamGear API Guide:

[**>>> Usage Guide**][camgear-wiki]

&nbsp;

&nbsp;


## VideoGear

> *VideoGear API provides a special internal wrapper around VidGear's exclusive [**Video Stabilizer**][stablizer-wiki] class.*

Furthermore, VideoGear API can provide internal access to both [CamGear](#camgear) and [PiGear](#pigear) APIs separated by a special flag. Thereby, _this API holds the exclusive power for any incoming VideoStream from any source, whether it is live or not, to access and stabilize it directly with minimum latency and memory requirements._

**Below is a snapshot of a VideoGear Stabilizer in action  (_See its detailed usage [here][stablizer-wiki-ex]_):**

<p align="center">
  <img src="https://github.com/abhiTronix/Imbakup/raw/master/Images/stabilizer.gif" alt="VideoGear Stabilizer in action!"/>
  <br>
  <sub><i>Original Video Courtesy <a href="http://liushuaicheng.org/SIGGRAPH2013/database.html" title="opensourced video samples database">@SIGGRAPH2013</a></i></sub>
</p>

**Code to generate above result:**

```python
# import required libraries
from vidgear.gears import VideoGear
import numpy as np
import cv2

# open any valid video stream with stabilization enabled(`stabilize = True`)
stream_stab = VideoGear(source = "test.mp4", stabilize = True).start()

# open same stream without stabilization for comparison
stream_org = VideoGear(source = "test.mp4").start()

# loop over
while True:

    # read stabilized frames
    frame_stab = stream_stab.read()

    # check for stabilized frame if Nonetype
    if frame_stab is None:
        break

    # read un-stabilized frame
    frame_org = stream_org.read()

    # concatenate both frames
    output_frame = np.concatenate((frame_org, frame_stab), axis=1)

    # put text over concatenated frame
    cv2.putText(
        output_frame, "Before", (10, output_frame.shape[0] - 10), cv2.FONT_HERSHEY_SIMPLEX,
        0.6, (0, 255, 0), 2,
    )
    cv2.putText(
        output_frame, "After", (output_frame.shape[1] // 2 + 10, output_frame.shape[0] - 10),
        cv2.FONT_HERSHEY_SIMPLEX,
        0.6, (0, 255, 0), 2,
    )

    # Show output window
    cv2.imshow("Stabilized Frame", output_frame)

    # check for 'q' key if pressed
    key = cv2.waitKey(1) & 0xFF
    if key == ord("q"):
        break

# close output window
cv2.destroyAllWindows()

# safely close both video streams
stream_org.stop()
stream_stab.stop()
```

### VideoGear API Guide:

[**>>> Usage Guide**][videogear-wiki]

&nbsp;

&nbsp;


## PiGear

> *PiGear is similar to CamGear but made to support various Raspberry Pi Camera Modules *(such as [OmniVision OV5647 Camera Module][OV5647-picam] and [Sony IMX219 Camera Module][IMX219-picam])*.*

PiGear provides a flexible multi-threaded wrapper around complete [`picamera`][picamera] python library to interface with these modules correctly, and also grants the ability to exploit its various parameters like `brightness, saturation, sensor_mode, etc.` effortlessly. 

Best of all, PiGear API provides excellent error-handling with features like a **Threaded Internal Timer** - that keeps active track of any frozen-threads/hardware-failures robustly, and exit safely if it does occur. So now if someone accidentally pulls Camera module cable out, when you're running PiGear API in your script, instead of going into possible kernel panic or frozen threads, this API will exit safely to save resources.

**Code to open picamera stream with variable parameters in PiGear API:**

```python
# import required libraries
from vidgear.gears import PiGear
import cv2

# add various Picamera tweak parameters to dictionary
options = {"hflip": True, "exposure_mode": "auto", "iso": 800, "exposure_compensation": 15, "awb_mode": "horizon", "sensor_mode": 0}

# open pi video stream with defined parameters
stream = PiGear(resolution = (640, 480), framerate = 60, logging = True, **options).start() 

# loop over
while True:

    # read frames from stream
    frame = stream.read()

    # check for frame if Nonetype
    if frame is None:
        break


    # {do something with the frame here}


    # Show output window
    cv2.imshow("Output Frame", frame)

    # check for 'q' key if pressed
    key = cv2.waitKey(1) & 0xFF
    if key == ord("q"):
        break

# close output window
cv2.destroyAllWindows()

# safely close video stream
stream.stop()
```
### PiGear API Guide:

[**>>> Usage Guide**][pigear-wiki]

&nbsp;

&nbsp;


## ScreenGear

> *ScreenGear API act as Screen Recorder, that can grab frames from your monitor in real-time either by define an area on the computer screen or fullscreen at the expense of insignificant latency. It also provide seamless support for capturing frames from multiple monitors.*

ScreenGear provides a high-level multi-threaded wrapper around [**python-mss**][mss] python library API and also supports a easy and flexible direct internal parameter manipulation. 

**Below is a snapshot of a ScreenGear API in action:**

<p align="center">
  <img src="https://github.com/abhiTronix/Imbakup/raw/master/Images/screengear.gif" alt="ScreenGear in action!" />
</p>

**Code to generate the above results:**

```python
# import required libraries
from vidgear.gears import ScreenGear
import cv2

# open video stream with default parameters
stream = ScreenGear().start()

# loop over
while True:

    # read frames from stream
    frame = stream.read()

    # check for frame if Nonetype
    if frame is None:
        break


    # {do something with the frame here}


    # Show output window
    cv2.imshow("Output Frame", frame)

    # check for 'q' key if pressed
    key = cv2.waitKey(1) & 0xFF
    if key == ord("q"):
        break

# close output window
cv2.destroyAllWindows()

# safely close video stream
stream.stop()
```

### ScreenGear API Guide:

[**>>> Usage Guide**][screengear-wiki]


&nbsp;

&nbsp;



## WriteGear

> *WriteGear handles various powerful Writer Tools that provide us the freedom to do almost anything imagine with multimedia files.*

WriteGear API provide a complete, flexible & robust wrapper around [**FFmpeg**][ffmpeg], a leading multimedia framework. With WriteGear, we can process real-time video frames into a lossless compressed format with any suitable specification in just few easy [lines of codes][compression-mode-ex]. These specifications include setting any video/audio property such as `bitrate, codec, framerate, resolution, subtitles,  etc.` easily as well complex tasks such as multiplexing video with audio in real-time(see this [example wiki][live-audio-wiki]). Best of all, WriteGear grants the freedom to play with any FFmpeg parameter with its exclusive custom Command function(see this [example wiki][custom-command-wiki]), while handling all errors robustly. 

In addition to this, WriteGear also provides flexible access to [**OpenCV's VideoWriter API**][opencv-writer] which provides some basic tools for video frames encoding but without compression.

**WriteGear primarily operates in the following two modes:**

  * **Compression Mode:** In this mode, WriteGear utilizes [**FFmpeg**][ffmpeg] inbuilt encoders to encode lossless multimedia files. It provides us the ability to exploit almost any available parameters available within FFmpeg, with so much ease and flexibility and while doing that it robustly handles all errors/warnings quietly. **You can find more about this mode [here][cm-writegear-wiki]**.


changelog.md
# CHANGELOG

## VidGear v0.1.7 <img src="https://raw.githubusercontent.com/abhiTronix/Imbakup/master/Images/new.gif" />

### New Features:

  * **WebGear API:**
    * Added a robust Live Video Server API that can transfer live video frames to any web browser on the network in real-time.
    * Implemented a flexible asyncio wrapper around [`starlette`](https://www.starlette.io/) ASGI Application Server.
    * Added seamless access to various starlette's Response classes, Routing tables, Static Files, Template engine(with Jinja2), etc.
    * Added a special internal access to VideoGear API and all its parameters.
    * Implemented a new Auto-Generation Work-flow to generate/download & thereby validate WebGear API data files from its GitHub server automatically.
    * Added on-the-go dictionary parameter in WebGear to tweak performance, Route Tables and other internal properties easily.
    * Added new simple & elegant default Bootstrap Cover Template for WebGear Server.
    * Added `__main__.py` to directly run WebGear Server through the terminal.
    * Added new gif and related docs for WebGear API.
    * Added and Updated various CI tests for this API.


  * **NetGear_Async API** 
    * Designed NetGear_Async asynchronous network API built upon ZeroMQ's asyncio API.
    * Implemented support for state-of-the-art asyncio event loop [`uvloop`](https://github.com/MagicStack/uvloop) at its backend.
    * Achieved Unmatchable high-speed and lag-free video streaming over the network with minimal resource constraint.
    * Added exclusive internal wrapper around VideoGear API for this API.
    * Implemented complete server-client handling and options to use variable protocols/patterns for this API.
    * Implemented support for  all four ZeroMQ messaging patterns: i.e `zmq.PAIR`, `zmq.REQ/zmq.REP`, `zmq.PUB/zmq.SUB`, and `zmq.PUSH/zmq.PULL`.
    * Implemented initial support for `tcp` and `ipc` protocols.
    * Added new Coverage CI tests for NetGear_Async Network Gear.
    * Added new Benchmark tests for benchmarking NetGear_Async against NetGear.

  * **Asynchronous Enhancements** 
    * Added `asyncio` package to vidgear for handling asynchronous APIs.
    * Moved WebGear API(webgear.py) to `asyncio` and created separate asyncio `helper.py` for it.
    * Various Performance tweaks for Asyncio APIs with concurrency within a single thread.
    * Moved `__main__.py` to asyncio for easier access to WebGear API through the terminal.
    * Updated `setup.py` with new dependencies and separated asyncio dependencies.

  * **General Enhancements:**
    
    * Added new highly-precise Threaded FPS class for accurate benchmarking with `time.perf_counter` python module.
    * Added a new [Gitter](https://gitter.im/vidgear/community) community channel.
    * Added a new *Reducer* function to reduce the frame size on-the-go.
    * Add *Flake8* tests to Travis CI to find undefined names. (PR by @cclauss)
    * Added a new unified `logging handler` helper function for vidgear.

### Updates/Improvements:
  
  * Re-implemented and simplified logic for NetGear Async server-end.
  * Added new dependencies for upcoming asyncio updates to `setup.py`.
  * Added `retry` function and replaced `wget` with `curl` for Linux test envs. 
  * Bumped OpenCV to latest `4.2.0-dev` for Linux test envs.
  * Updated YAML files to reflect new changes to different CI envs.
  * Separated each API logger with a common helper method to avoid multiple copies. 
  * Limited Importing OpenCV API version check's scope to `helper.py` only.
  * Implemented case for incorrect `color_space` value in ScreenGear API.
  * Removed old conflicting logging formatter with a common method and expanded logging.
  * Improved and added `shutdown` function for safely stopping frame producer threads in WebGear API.
  * Re-implemented and simplified all CI tests with maximum code-coverage in mind.
  * Replaced old `mkdir` function with new `mkdir_safe` helper function for creating directories safely.
  * Updated ReadMe.md with updated diagrams, gifs and information.
  * Improve, structured and Simplified the Contribution Guidelines.
  * Bundled CI requirements in a single command.(Suggested by @cclauss)
  * Replaced line endings CRLF with LF endings.
  * Added dos2unix for Travis OSX envs.
  * Bumped Codecov to maximum. 

### Breaking Updates / Improvements / Changes

  * ***:warning: Dropped support for Python 3.5 and below legacies. (See [issue #99](https://github.com/abhiTronix/vidgear/issues/99))***
  * Dropped and replaced Python 3.5 matrices with new Python 3.8 matrices in all CI environments.
  * Implemented PEP-8 Styled [**Black**](https://github.com/psf/black) formatting throughout the source-code.
  * Limited protocols support to `tcp` and `ipc` only, in NetGear API.

### Fixes:
  
  * Fixed Major NetGear_Async bug where `__address` and `__port` are not set in async mode.(PR by @otter-in-a-suit) 
  * Fixed Major PiGear Color-space Conversion logic bug.
  * Workaround for CAP_IMAGES: error in YouTube Mode.
  * Replaced incorrect `terminate()` with `join()` in PiGear.
  * Removed `uvloop` for windows as still [NOT yet supported](https://github.com/MagicStack/uvloop/issues/14).
  * Refactored Asynchronous Package name `async` to `asyncio`, since it is used as Keyword in python>=3.7*(raises SyntaxError)*.
  * Fixed unfinished close of event loops bug in WebGear API.
  * Fixed NameError in helper.py.
  * Added fix for OpenCV installer failure on Linux test envs.
  * Fixed undefined NameError in `helper.py` context. (@cclauss)
  * Fixed incorrect logic while pulling frames from ScreenGear API.
  * Fixed missing functions in `__main__.py`.
  * Fixed Typos and definitions in docs.
  * Added missing `camera_num` parameter to VideoGear.
  * Added OpenSSL's [SSL: CERTIFICATE_VERIFY_FAILED] bug workaround for macOS envs.
  * Removed `download_url` meta from setup.py.

vidgear/gears/webgear.py
"""
===============================================
vidgear library source-code is deployed under the Apache 2.0 License:

Copyright (c) 2019 Abhishek Thakur(@abhiTronix) <abhi.una12@gmail.com>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
===============================================
"""

# import the necessary packages
from starlette.applications import Starlette
from starlette.responses import StreamingResponse
from starlette.templating import Jinja2Templates
from starlette.staticfiles import StaticFiles
from starlette.routing import Mount
from starlette.routing import Route

from .videogear import VideoGear
from .helper import reducer
from .helper import logger_handler
from .helper import generate_webdata
from collections import deque

import logging as log
import os, cv2, asyncio, sys


#define logger
logger = log.getLogger('WebGear')
logger.addHandler(logger_handler())
logger.setLevel(log.DEBUG)



class WebGear:

	"""
	WebGear is a powerful ASGI Video-streamer API, that is built upon Starlette - a lightweight ASGI framework/toolkit, which is ideal 
	for building high-performance asyncio services.

	WebGear API provides a flexible but robust asyncio wrapper around Starlette ASGI application and can easily access its various components 
	independently. Thereby providing it the ability to interact with the Starlette's ecosystem of shared middleware and mountable applications 
	& seamless access to its various Response classes, Routing tables, Static Files, Templating engine(with Jinja2), etc.

	WebGear acts as robust Live Video Streaming Server that can stream live video frames to any web browser on a network in real-time. 
	It addition to this, WebGear provides a special internal wrapper around VideoGear API, which itself provides internal access to both CamGear and PiGear 
	APIs thereby granting it exclusive power for streaming frames incoming from any device/source. Also on the plus side, since WebGear has access 
	to all functions of VideoGear API, therefore it can stabilize video frames even while streaming live.


	WebGear specific parameters:

    - ** options:(dict): can be used in addition, to pass user-defined parameter to WebGear API in the form of python dictionary. Supported WebGear 
    					dictionary attributes are:

        - `custom_data_location` (str): Can be used to change/alter `default location` path to somewhere else. 

        - `overwrite_default_files` (bool): Can be used to force trigger the `Auto-generation process` to overwrite existing data-files. Remember only downloaded 
        									files will be overwritten in this process, and any other file/folder will NOT be affected/overwritten.

        - `frame_size_reduction`: (int/float) This attribute controls the size reduction (in percentage) of the frame to be streamed on Server. 
        										Its value can be between 0-90, and the recommended value is 40. 

        - JPEG Encoding Parameters: In WebGear, the input video frames are first encoded into Motion JPEG (M-JPEG or MJPEG) video compression format in which 
        						each video frame or interlaced field of a digital video sequence is compressed separately as a JPEG image, before sending onto 
        						a server. Therefore, WebGear API provides various attributes to have full control over JPEG encoding performance and quality, 
        						which are as follows:

            - `frame_jpeg_quality`(int) It controls the JPEG encoder quality and value varies from 0 to 100 (the higher is the better quality but 
            								performance will be lower). Its default value is 95. 

            - `frame_jpeg_optimize`(bool) It enables various JPEG compression optimizations such as Chroma subsampling, Quantization table, etc. 
            							Its default value is `False`. 

            - `frame_jpeg_progressive`(bool) It enables Progressive JPEG encoding instead of the Baseline. Progressive Mode. 
            								Its default value is `False` means baseline mode. 


	VideoGear Specific parameters for WebGear:
	
		:param (boolean) enablePiCamera: set this flag to access PiGear or CamGear class respectively. 
										/ This means the if enablePiCamera flag is `True`, PiGear class will be accessed 
										/ and if `False`, the camGear Class will be accessed. Its default value is False.

		:param (boolean) stabilize: set this flag to enable access to VidGear's Stabilizer Class. This basically enables(if True) or disables(if False) 
										video stabilization in VidGear. Its default value is False.

		:param (dict) **options: can be used in addition, to pass parameter supported by VidGear's stabilizer class.
								/ Supported dict keys are: 
									- `SMOOTHING_RADIUS` (int) : to alter averaging window size. It handles the quality of stabilization at expense of latency and sudden panning. 
															/ Larger its value, less will be panning, more will be latency and vice-versa. It's default value is 30.
									- `BORDER_SIZE` (int) : to alter output border cropping. It's will crops the border to reduce the black borders from stabilization being too noticeable. 
															/ Larger its value, more will be cropping. It's default value is 0 (i.e. no cropping).			
									- `BORDER_TYPE` (string) : to change the border mode. Valid border types are 'black', 'reflect', 'reflect_101', 'replicate' and 'wrap'. It's default value is 'black'
	

	CamGear Specific supported parameters for WebGear:

		:param source : take the source value for CamGear Class. Its default value is 0. Valid Inputs are:
			- Index(integer): Valid index of the video device.
			- YouTube Url(string): Youtube URL as input.
			- Network_Stream_Address(string): Incoming Stream Valid Network address. 
			- GStreamer (string) videostream Support
		:param (boolean) y_tube: enables YouTube Mode in CamGear Class, i.e If enabled the class will interpret the given source string as YouTube URL. 
								/ Its default value is False.
		:param (int) backend: set the backend of the video stream (if specified). Its default value is 0.


	PiGear Specific supported parameters for WebGear:
		:param (integer) camera_num: selects the camera module index that will be used by API. 
								/	Its default value is 0 and shouldn't be altered until unless 
								/	if you using Raspberry Pi 3/3+ compute module in your project along with multiple camera modules. 
								/	Furthermore, Its value can only be greater than zero, otherwise, it will throw ValueError for any negative value.
		:param (tuple) resolution: sets the resolution (width,height) in Picamera class. Its default value is (640,480).
		:param (integer) framerate: sets the framerate in Picamera class. Its default value is 25.


	Common parameters for WebGear: 
		:param (string) colorspace: set colorspace of the video stream. Its default value is None.
		:param (dict) **options: parameter supported by various API (whichever being accessed).
		:param (boolean) logging: set this flag to enable/disable error logging essential for debugging. Its default value is False.
		:param (integer) time_delay: sets time delay(in seconds) before start reading the frames. 
							/ This delay is essentially required for camera to warm-up. 
							/ Its default value is 0.
	"""


	def __init__(self, enablePiCamera = False, stabilize = False, source = 0, camera_num = 0, y_tube = False, backend = 0, colorspace = None, resolution = (640, 480), framerate = 25, logging = False, time_delay = 0, **options):

		#initialize global params
		self.__jpeg_quality = 90 #90% quality
		self.__jpeg_optimize = 0 #optimization off
		self.__jpeg_progressive=0 #jpeg will be baseline instead
		self.__frame_size_reduction = 20 #20% reduction
		self.__logging = logging

		custom_data_location = '' #path to save data-files to custom location
		data_path = '' #path to WebGear data-files
		overwrite_default=False

		#reformat dictionary
		options = {str(k).strip(): v for k,v in options.items()}

		# assign values to global variables if specified and valid
		if options:
			if 'frame_size_reduction' in options:
				value = options["frame_size_reduction"]
				if isinstance(value, (int, float)) and value >= 0 and value <= 90:
					self.__frame_size_reduction = value
				else: 
					logger.warning("Skipped invalid `frame_size_reduction` value!")
				del options['frame_size_reduction'] #clean
			if 'frame_jpeg_quality' in options:
				value = options["frame_jpeg_quality"]
				if isinstance(value, (int, float)) and value >= 10 and value <= 95:
					self.__jpeg_quality = int(value)
				else: 
					logger.warning("Skipped invalid `frame_jpeg_quality` value!")
				del options['frame_jpeg_quality'] #clean
			if 'frame_jpeg_optimize' in options:
				value = options["frame_jpeg_optimize"]
				if isinstance(value, bool): 
					self.__jpeg_optimize = int(value)
				else: 
					logger.warning("Skipped invalid `frame_jpeg_optimize` value!")
				del options['frame_jpeg_optimize'] #clean
			if 'frame_jpeg_progressive' in options:
				value = options["frame_jpeg_progressive"]
				if isinstance(value, bool):
					self.__jpeg_progressive = int(value)
				else: 
					logger.warning("Skipped invalid `frame_jpeg_progressive` value!")
				del options['frame_jpeg_progressive'] #clean
			if 'custom_data_location' in options:
				value = options["custom_data_location"]
				if isinstance(value,str):
					assert os.access(value, os.W_OK), "[WebGear:ERROR] :: Permission Denied!, cannot write WebGear data-files to '{}' directory!".format(value)
					assert os.path.isdir(os.path.abspath(value)), "[WebGear:ERROR] :: `custom_data_location` value must be the path to a directory and not to a file!"
					custom_data_location = os.path.abspath(value)
				else:
					logger.warning("Skipped invalid `custom_data_location` value!")
				del options['custom_data_location'] #clean
			if 'overwrite_default_files' in options:
				value = options["overwrite_default_files"]
				if isinstance(value, bool):
					overwrite_default = value
				else:
					logger.warning("Skipped invalid `overwrite_default_files` value!")
				del options['overwrite_default_files'] #clean

		#define stream with necessary params
		self.stream = VideoGear(enablePiCamera = enablePiCamera, stabilize = stabilize, source = source, camera_num = camera_num, y_tube = y_tube, backend = backend, colorspace = colorspace, resolution = resolution, framerate = framerate, logging = logging, time_delay = time_delay, **options)

		#check if custom certificates path is specified
		if custom_data_location:
			data_path = generate_webdata(custom_data_location, overwrite_default = overwrite_default, logging = logging)
		else:
			# otherwise generate suitable path
			from os.path import expanduser
			data_path = generate_webdata(os.path.join(expanduser("~"),".vidgear"), overwrite_default = overwrite_default, logging = logging)
		
		#log it
		if self.__logging: logger.debug('`{}` is the default location for saving WebGear data-files.'.format(data_path))
		if self.__logging: logger.debug('Setting params:: Size Reduction:{}%, JPEG quality:{}%, JPEG optimizations:{}, JPEG progressive:{}'.format(self.__frame_size_reduction, self.__jpeg_quality, bool(self.__jpeg_optimize), bool(self.__jpeg_progressive)))

		#define Jinja2 templates handler
		self.__templates = Jinja2Templates(directory='{}/templates'.format(data_path))

		#define custom exception handlers
		self.__exception_handlers = {404: self.__not_found,
									500: self.__server_error}
		#define routing tables
		self.routes = [Route('/', endpoint=self.__homepage),
						Route('/video', endpoint=self.__video),
						Mount('/static', app=StaticFiles(directory='{}/static'.format(data_path)), name="static")]
		#copying original routing tables for further validation
		self.__rt_org_copy = self.routes[:]
		#keeps check if producer loop should be running
		self.__isrunning = True



	def __call__(self):
		"""
		Implements custom callable method
		"""
		#validate routing tables
		assert not(self.routes is None), "Routing tables are NoneType!"
		if not isinstance(self.routes, list) or not all(x in self.routes for x in self.__rt_org_copy): raise RuntimeError("Routing tables are not valid!")
		#initiate stream
		if self.__logging: logger.debug('Initiating Video Streaming.')
		self.stream.start()
		#return Starlette application
		if self.__logging: logger.debug('Running Starlette application.')
		return Starlette(debug = (True if self.__logging else False), routes=self.routes, exception_handlers=self.__exception_handlers, on_shutdown=[self.shutdown])



	async def __producer(self):
		"""
		Implements async frame producer.
		"""
		# loop over frames
		while self.__isrunning:
			#read frame
			frame = self.stream.read()
			#break if NoneType
			if frame is None: break
			#reducer frames size if specified
			if self.__frame_size_reduction: frame = reducer(frame, percentage = self.__frame_size_reduction)
			#handle JPEG encoding
			encodedImage = cv2.imencode('.jpg', frame, [cv2.IMWRITE_JPEG_QUALITY, self.__jpeg_quality, cv2.IMWRITE_JPEG_PROGRESSIVE, self.__jpeg_progressive,cv2.IMWRITE_JPEG_OPTIMIZE, self.__jpeg_optimize])[1].tobytes()
			#yield frame in byte format
			yield  (b'--frame\r\nContent-Type:image/jpeg\r\n\r\n'+encodedImage+b'\r\n')
			await asyncio.sleep(0.01)



	async def __video(self, scope):
		"""
		Return a async video streaming response.
		"""
		assert scope['type'] == 'http'
		return StreamingResponse(self.__producer(), media_type='multipart/x-mixed-replace; boundary=frame')



	async def __homepage(self, request):
		"""
		Return an HTML index page.
		"""
		return self.__templates.TemplateResponse('index.html', {'request': request})



	async def __not_found(self, request, exc):
		"""
		Return an HTML 404 page.
		"""
		return self.__templates.TemplateResponse('404.html', {'request': request}, status_code=404)



	async def __server_error(self, request, exc):
		"""
		Return an HTML 500 page.
		"""
		return self.__templates.TemplateResponse('500.html', {'request': request}, status_code=500)



	def shutdown(self):
		"""
		Implements a callable to run on application shutdown
		"""
		if not(self.stream is None):
			if self.__logging: logger.debug('Closing Video Streaming.')
			#stops producer
			self.__isrunning = False
			#stops VideoGear stream
			self.stream.stop()
			#prevent any re-iteration
			self.stream = None